File: backend/src/test/java/lan/dk/podcastserver/manager/downloader/RTMPDownloaderTest.java
package lan.dk.podcastserver.manager.downloader;

import com.github.davinkevin.podcastserver.service.MimeTypeService;
import com.github.davinkevin.podcastserver.service.ProcessService;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.entity.Status;
import lan.dk.podcastserver.manager.ItemDownloadManager;
import lan.dk.podcastserver.repository.ItemRepository;
import lan.dk.podcastserver.repository.PodcastRepository;
import lan.dk.podcastserver.service.properties.ExternalTools;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import com.github.davinkevin.podcastserver.IOUtils;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.util.FileSystemUtils;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.util.Optional;
import java.util.UUID;

import static io.vavr.API.List;
import static io.vavr.API.Try;
import static com.github.davinkevin.podcastserver.IOUtils.ROOT_TEST_PATH;
import static com.github.davinkevin.podcastserver.IOUtils.TEMPORARY_EXTENSION;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyVararg;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.*;

/**
 * Created by kevin on 27/03/2016 for Podcast Server
 */
@RunWith(MockitoJUnitRunner.class)
public class RTMPDownloaderTest {

    private @Mock ExternalTools externalTools;
    private @Mock PodcastRepository podcastRepository;
    private @Mock ItemRepository itemRepository;
    private @Mock ItemDownloadManager itemDownloadManager;
    private @Mock PodcastServerParameters podcastServerParameters;
    private @Mock SimpMessagingTemplate template;
    private @Mock MimeTypeService mimeTypeService;
    private @Mock ProcessService processService;
    private @InjectMocks RTMPDownloader rtmpDownloader;

    private @Captor ArgumentCaptor<String> processParameters;

    private Item item;

    @Before
    public void beforeEach() {
        Podcast podcast = Podcast.builder()
                .id(UUID.randomUUID())
                .title("RTMP Podcast")
                .build();
        item = Item
                .builder()
                .url("rtmp://a.url.com/foo/bar.mp4")
                .status(Status.STARTED)
                .podcast(podcast)
                .progression(0)
                .numberOfFail(0)
                .build();
        when(podcastServerParameters.getDownloadExtension()).thenReturn(TEMPORARY_EXTENSION);
        when(externalTools.getRtmpdump()).thenReturn("/usr/local/bin/rtmpdump");
        when(podcastServerParameters.getRootfolder()).thenReturn(ROOT_TEST_PATH);
        when(podcastRepository.findById(eq(podcast.getId()))).thenReturn(Optional.of(podcast));

        rtmpDownloader.setDownloadingItem(DownloadingItem.builder().item(item).urls(List()).build());
        rtmpDownloader.setItemDownloadManager(itemDownloadManager);

        rtmpDownloader.postConstruct();
        FileSystemUtils.deleteRecursively(ROOT_TEST_PATH.resolve(podcast.getTitle()).toFile());
        Try(() -> Files.createDirectories(ROOT_TEST_PATH));
    }

    @Test
    public void should_be_a_downloader() {
        assertThat(Downloader.class.isInstance(rtmpDownloader));
        assertThat(AbstractDownloader.class.isInstance(rtmpDownloader));
    }

    @Test
    public void should_download() throws IOException, URISyntaxException {
        /* Given */
        ProcessBuilder processBuilder = new ProcessBuilder("/bin/cat", IOUtils.toPath("/remote/downloader/rtmpdump/rtmpdump.txt").get().toString());
        when(processService.newProcessBuilder((String[]) anyVararg())).then(i -> {
            Files.createFile(ROOT_TEST_PATH.resolve(item.getPodcast().getTitle()).resolve("bar.mp4" + TEMPORARY_EXTENSION));
            return processBuilder;
        });
        when(processService.pidOf(any())).thenReturn(1234);

        /* When */
        rtmpDownloader.download();

        /* Then */
        assertThat(rtmpDownloader.pid).isEqualTo(0);
        assertThat(ROOT_TEST_PATH.resolve(item.getPodcast().getTitle()).resolve("bar.mp4")).exists();
        verify(processService, times(1)).newProcessBuilder(processParameters.capture());
        assertThat(processParameters.getAllValues())
                .hasSize(5)
                .containsExactly(
                        "/usr/local/bin/rtmpdump",
                        "-r",
                        item.getUrl(),
                        "-o",
                        ROOT_TEST_PATH.resolve(item.getPodcast().getTitle()).resolve("bar.mp4" + TEMPORARY_EXTENSION).toFile().getAbsolutePath()
                );

    }

    @Test
    public void should_stop_download_if_ioexception() throws URISyntaxException {
        /* Given */
        ProcessBuilder processBuilder = new ProcessBuilder("/bin");
        when(processService.newProcessBuilder((String[]) anyVararg())).then(i -> processBuilder);
        rtmpDownloader.p = mock(Process.class);

        /* When */
        assertThatThrownBy(() -> rtmpDownloader.download())

        /* Then */
                .isInstanceOf(RuntimeException.class);
    }

    @Test
    public void should_stop_process_if_failed() {
        /* GIVEN */
        ProcessBuilder processBuilder = new ProcessBuilder("/bin");
        rtmpDownloader.p = mock(Process.class);
        /* WHEN  */
        rtmpDownloader.failDownload();
        /* THEN  */
        verify(rtmpDownloader.p, times(1)).destroy();
        assertThat(rtmpDownloader.item.getStatus()).isEqualByComparingTo(Status.FAILED);
    }

    @Test
    public void should_destroy_previous_process_at_start() throws URISyntaxException {
        /* Given */
        Process process = mock(Process.class);
        rtmpDownloader.pid = 123;
        rtmpDownloader.p = process;

        ProcessBuilder processBuilder = new ProcessBuilder("/bin/cat", IOUtils.toPath("/remote/downloader/rtmpdump/rtmpdump.txt").toString());
        when(processService.newProcessBuilder((String[]) anyVararg())).then(i -> {
            Files.createFile(ROOT_TEST_PATH.resolve(item.getPodcast().getTitle()).resolve("bar.mp4" + TEMPORARY_EXTENSION));
            return processBuilder;
        });
        when(processService.pidOf(any())).thenReturn(1234);

        /* When */
        rtmpDownloader.startDownload();

        /* Then */
        verify(process, times(1)).destroy();
    }

    @Test
    public void should_pause_process_on_pause_of_download() {
        /* Given */
        rtmpDownloader.pid = 123;
        when(processService.newProcessBuilder((String[]) anyVararg())).thenReturn(new ProcessBuilder("/bin/ls"));

        /* When */
        rtmpDownloader.pauseDownload();

        /* Then */
        verify(processService, times(1)).newProcessBuilder(processParameters.capture());
        assertThat(processParameters.getAllValues())
                .hasSize(3)
                .containsExactly(
                        "kill",
                        "-STOP",
                        "123"
                );
    }

    @Test
    public void should_stop_if_pause_not_working() {
        /* Given */
        Process process = mock(Process.class);
        rtmpDownloader.pid = 123;
        rtmpDownloader.p = process;
        when(processService.newProcessBuilder((String[]) anyVararg())).thenReturn(new ProcessBuilder("/bin"));

        /* When */
        rtmpDownloader.pauseDownload();

        /* Then */
        verify(process, times(1)).destroy();
        assertThat(item.getStatus()).isEqualTo(Status.STOPPED);
    }

    @Test
    public void should_be_compatible() {
        assertThat(rtmpDownloader.compatibility(DownloadingItem.builder().urls(List("http://foo.bar.com/video")).build())).isGreaterThan(1);
        assertThat(rtmpDownloader.compatibility(DownloadingItem.builder().urls(List("rtmp://foo.bar.com/video")).build())).isEqualTo(1);
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/download/downloaders/rtmp/RTMPDownloaderTest.kt
package com.github.davinkevin.podcastserver.download.downloaders.rtmp

import com.github.davinkevin.podcastserver.download.downloaders.DownloaderHelper
import com.github.davinkevin.podcastserver.download.downloaders.DownloaderHelperFactory
import com.github.davinkevin.podcastserver.download.downloaders.DownloadingInformation
import com.github.davinkevin.podcastserver.download.downloaders.DownloadingItem
import com.github.davinkevin.podcastserver.entity.Status
import com.github.davinkevin.podcastserver.service.ProcessService
import com.github.davinkevin.podcastserver.service.properties.ExternalTools
import com.github.davinkevin.podcastserver.service.storage.FileMetaData
import com.github.davinkevin.podcastserver.service.storage.UploadRequest
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito
import org.mockito.kotlin.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Import
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.io.ByteArrayInputStream
import java.io.File
import java.net.URI
import java.nio.file.Files
import java.nio.file.Paths
import java.time.Clock
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.util.*
import java.util.concurrent.CompletableFuture.runAsync
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeUnit.SECONDS
import kotlin.io.path.Path

private val fixedDate = OffsetDateTime.of(2019, 3, 4, 5, 6, 7, 0, ZoneOffset.UTC)

/**
 * Created by kevin on 27/03/2016 for Podcast Server
 */
@ExtendWith(SpringExtension::class)
class RTMPDownloaderTest(
    @Autowired private val downloader: RTMPDownloader,
    @Autowired private val helper: DownloaderHelper,
) {

    @MockitoBean lateinit var externalTools: ExternalTools
    @MockitoBean lateinit var processService: ProcessService

    private val item: DownloadingItem = DownloadingItem (
            id = UUID.randomUUID(),
            title = "Title",
            status = Status.NOT_DOWNLOADED,
            url = URI("http://a.fake.url/with/file.mp4?param=1"),
            numberOfFail = 0,
            progression = 0,
            podcast = DownloadingItem.Podcast(
                    id = UUID.randomUUID(),
                    title = "A Fake ffmpeg Podcast"
            ),
            cover = DownloadingItem.Cover(
                    id = UUID.randomUUID(),
                    url = URI("https://bar/foo/cover.jpg")
            )
    )

    @TestConfiguration
    @Import(RTMPDownloader::class)
    class LocalTestConfiguration {
        @Bean fun helper(): DownloaderHelper = DownloaderHelperFactory(
            downloadRepository = mock(),
            clock = Clock.fixed(fixedDate.toInstant(), ZoneId.of("UTC")),
            file = mock(),
            template = mock(),
        ).build(mock(), mock())
    }

    @Nested
    inner class DownloadTest {

        lateinit var executionLogs: ByteArrayInputStream

        @BeforeEach
        fun beforeEach() {
            executionLogs = """
            Line with no information usable
            Progression : (1%)
            Progression : (2%)
            Progression : (3%)
            Download Complete
        """
                    .trimIndent()
                    .toByteArray()
                    .inputStream()

            whenever(externalTools.rtmpdump).thenReturn("/usr/local/bin/rtmpdump")
            helper.info = DownloadingInformation(item, listOf(), Path("file.mp4"), null)
        }

        @Nested
        @DisplayName("should failed ")
        inner class ShouldFailed {

            @Test
            fun `immediately when process start`() {
                /* Given */
                val pb = mock<ProcessBuilder>()
                whenever(pb.directory(File("/tmp"))).then { pb }
                whenever(pb.redirectErrorStream(true)).then { pb }
                whenever(processService.newProcessBuilder(
                    eq("/usr/local/bin/rtmpdump"),
                    eq("-r"),
                    eq(item.url.toASCIIString()),
                    eq("-o"),
                    any<String>(),
                )).then { pb }
                whenever(processService.start(any())).then { throw RuntimeException("Error occur during shell execution")}
                whenever(helper.downloadRepository.updateDownloadItem(any())).thenReturn(0)

                /* When */
                downloader.run()

                /* Then */
                assertThat(downloader.downloadingInformation.item.status).isEqualTo(Status.FAILED)
            }

            @Test
            fun after_launch_process() {
                /* Given */
                val p = mock<Process>()
                val pb = mock<ProcessBuilder>()
                whenever(pb.directory(File("/tmp"))).then { pb }
                whenever(pb.redirectErrorStream(true)).then { pb }
                whenever(processService.newProcessBuilder(
                    eq("/usr/local/bin/rtmpdump"),
                    eq("-r"),
                    eq(item.url.toASCIIString()),
                    eq("-o"),
                    any<String>(),
                )).then { pb }
                whenever(processService.start(any())).then { p }
                whenever(processService.pidOf(p)).then { throw RuntimeException("Error during pid fetching") }
                whenever(helper.downloadRepository.updateDownloadItem(any())).thenReturn(0)

                /* When */
                downloader.run()

                /* Then */
                assertThat(downloader.downloadingInformation.item.status).isEqualTo(Status.FAILED)
                verify(p).destroy()
            }
        }

        @Nested
        @DisplayName("should download")
        inner class ShouldDownload {

            private val pb = mock<ProcessBuilder>()
            private val p = mock<Process>()
            private val pid = 1234L

            @BeforeEach
            fun beforeEach() {
                var fileToCreate = ""
                Mockito.reset(processService, p, pb)

                whenever(pb.directory(File("/tmp"))).then { pb }
                whenever(pb.redirectErrorStream(true)).then { pb }
                whenever(processService.newProcessBuilder(
                    eq("/usr/local/bin/rtmpdump"),
                    eq("-r"),
                    eq(item.url.toASCIIString()),
                    eq("-o"),
                    any<String>(),
                )).then {
                    fileToCreate = it.arguments[4]!! as String
                    pb
                }
                whenever(processService.start(pb)).then {
                    Files.createFile(Paths.get(fileToCreate))
                    p
                }
                whenever(processService.pidOf(any())).thenReturn(pid)
                whenever(p.inputStream).then { executionLogs }
            }

            @Test
            fun `and save file to disk`() {
                /* Given */
                doNothing().whenever(helper.file)
                    .upload(argThat<UploadRequest.ForItemFromPath> { podcastTitle == item.podcast.title })
                whenever(helper.file.metadata(eq(item.podcast.title), any()))
                    .thenReturn(FileMetaData("video/mp4", 123L))
                whenever(helper.downloadRepository.updateDownloadItem(any())).thenReturn(0)
                whenever(helper.downloadRepository.finishDownload(
                        id = item.id,
                        length = 123L,
                        mimeType = "video/mp4",
                        fileName = Path("file-${item.id}.mp4"),
                        downloadDate = fixedDate
                )).thenReturn(0)

                /* When */
                downloader.run()

                /* Then */
                await().atMost(5, TimeUnit.SECONDS).untilAsserted {
                    assertThat(downloader.downloadingInformation.item.status).isEqualTo(Status.FINISH)
                }
            }

            @Nested
            @DisplayName("and then")
            inner class AndThen {

                private var isWaiting = false

                @BeforeEach
                fun beforeEach() {
                    whenever(p.waitFor()).then { isWaiting = true; SECONDS.sleep(4); 0 }
                    whenever(helper.downloadRepository.updateDownloadItem(any())).thenReturn(0)
                }

                @Test
                fun stop() {
                    /* GIVEN */
                    doNothing().whenever(helper.file)
                        .upload(argThat<UploadRequest.ForItemFromPath> { podcastTitle == item.podcast.title })
                    whenever(helper.file.metadata(eq(item.podcast.title), any()))
                        .thenReturn(FileMetaData("video/mp4", 123L))
                    whenever(helper.downloadRepository.finishDownload(
                            id = item.id,
                            length = 123L,
                            mimeType = "video/mp4",
                            fileName = Path("file-${item.id}.mp4"),
                            downloadDate = fixedDate
                    )).thenReturn(0)

                    /* WHEN  */
                    runAsync { downloader.run() }
                    await().until { isWaiting }
                    downloader.stopDownload()

                    /* THEN  */
                    await().atMost(5, TimeUnit.SECONDS).untilAsserted {
                        assertThat(downloader.downloadingInformation.item.status).isEqualTo(Status.STOPPED)
                        verify(p).destroy()
                    }
                }

            }

            @Test
            fun `and handle unexpected end of download logs`() {
                /* Given */
                val executionLogs = """
                Line with no information usable
                Progression : (1%)
                Progression : (2%)
                Progression : (3%)
                """
                        .trimIndent()
                        .toByteArray()
                        .inputStream()
                whenever(helper.downloadRepository.updateDownloadItem(any())).thenReturn(0)
                whenever(p.inputStream).then { executionLogs }

                /* When */
                downloader.run()

                /* Then */
                assertThat(downloader.downloadingInformation.item.status).isEqualTo(Status.FAILED)
            }
        }
    }

}
